cmd/ebrelayer/relayer/ethereum.go

```diff
@@ -171,26 +182,51 @@ func (sub EthereumSub) Start(completionEvent *sync.WaitGroup) {
 			completionEvent.Add(1)
 			go sub.Start(completionEvent)
 			return
+		case err := <-subHead.Err():
+			sub.Logger.Error(err.Error())
+			completionEvent.Add(1)
+			go sub.Start(completionEvent)
+			return
+		case newHead := <-heads:
+			sub.Logger.Info(fmt.Sprintf("New header %d with hash %v", newHead.Number, newHead.Hash()))
+
+			events, ok := sub.EventsBuffer.Buffer[newHead.Number]
+			if ok && events.BlockHash != newHead.Hash() {
+				// Deal with block reorg, delete all events with old hash value
+				sub.Logger.Info(fmt.Sprintf("Block reorg found old hash is %v new hash is %v", events.BlockHash, newHead.Hash()))
+				delete(sub.EventsBuffer.Buffer, newHead.Number)
+			}
+
+			// Iterate find out older enough events
+			for key, value := range sub.EventsBuffer.Buffer {
+				sub.Logger.Info(fmt.Sprintf("New header is %d stored header is %d", newHead.Number, key))
```

Comment:
I just realized that we have a logic error here. This picks up events and stores them in the q, but it doesn't check that this event is still valid and that TX did not end up getting dis included in a block because the block was orphaned, which was the problem we are trying to solve.

Here, we will need to check on the transaction that was included in this event and make sure that it is in fact in a block that is 30 blocks ago, if it is not in a previous block, we delete it from the q, if it got included in a later block, we add it again to the queue.

Reply:
I believe the we can check it via code logic before. If we get the event, which block hash not the same with the buffered one. Then we removed all events with old block hash.
			if ok && events.BlockHash != newHead.Hash() {
				// Deal with block reorg, delete all events with old hash value
				sub.Logger.Info(fmt.Sprintf("Block reorg found old hash is %v new hash is %v", events.BlockHash, newHead.Hash()))
				delete(sub.EventsBuffer.Buffer, newHead.Number)
			}

Reply:
That solution sounds good to me.

Reply:
We have to somehow query to check if this old block was valid, if it is invalid, like you said, we throw all of the old stuff out.